Add support for TI CC13x2 and CC26x2 SoCs

bwitherspoon opened the pull request:
I am working on adding support for the TI SimpleLink CC13x2 and CC26x2 series of SoCs. This WIP PR is to track my progress and get early comments if possible.

So far I have implemented pinmux, GPIO, UART, and TRNG drivers. I am currently working on a IEEE 802.15.4 driver and then plan to add drivers for other peripherals starting with SPI and the WDT. I have used the low-level HAL (driverlib) from TI. I am calling these the simplelink drivers and intend on them supporting other SoCs in the SimpleLink family if also supported by the driverlib HAL. I am also working with the CC2650 SoC.

This PR also includes support for the TI CC26x2R and CC13x2R LaunchPad development boards.

bwitherspoon committed:
ext: simplelink: import HAL from TI CC13x2 / CC26x2 SDK 3.10.00.53

Add low level drivers (driverlib) and RF patches from the TI CC13x2 and
CC26x2 SDK. These sources have been added unmodified with the exception
of converting DOS to UNIX line endings.

The majority of these APIs are available in ROM and inclusion of the
headers will map functions that are not inline to the ROM versions.

Origin: Texas Instruments SimpleLink CC13x2 and CC26x2 SDK
License: BSD 3-Clause
URL: http://www.ti.com/tool/simplelink-cc13x2-26x2-sdk
Purpose: Provides HAL for TI CC13x2 and CC26x2 SoCs
Maintained-by: External

Signed-off-by: Brett Witherspoon <spoonb@cdspooner.com>

bwitherspoon committed:
boards: cc26x2r1_launchpad: add support for TI CC26x2R LaunchPad

The TI CC26x2R LaunchPad (LAUNCHXL-CC26X2R1) is a development kit that
features the CC2652R SoC.

Signed-off-by: Brett Witherspoon <spoonb@cdspooner.com>

bwitherspoon committed:
boards: cc1352r1_launchxl: add support for TI CC1352R LaunchPad

The TI CC1352R LaunchPad (LAUNCHXL-CC1352R1) is a development kit that
features the CC1352R SoC.

Signed-off-by: Brett Witherspoon <spoonb@cdspooner.com>

bwitherspoon commented:
There are some sources from the TI SDK (ext/hal/ti/simplelink/source/ti/devices/cc13x2_cc26x2/driverlib/sw_*) which are either not licensed or public domain. The TI SDK manifest says the Core SDK is BSD 3-clause. However, these sources are not used so they could potentially be removed.

bwitherspoon commented:
I am missing images for the boards.

n3rd4i commented:
Will this work support either:
1. CC26x2 BLE stack?
2. maybe even CC26x0 in the future (with the same question as above)?

Matt90o commented:
Great start, but am indeed wondering if the BLE stack can be run on the board and if it would fit. Maybe TI's micro BLE stack would fit using GCC? http://dev.ti.com/tirex/content/simplelink_cc2640r2_sdk_1_30_00_25/docs/blestack/ble_sw_dev_guide/html/u-stack/index.html

dbkinder commented on boards/arm/cc1352r_launchpad/doc/index.rst:
```diff
@@ -0,0 +1,177 @@
+*************************
+
+Before flashing or debugging ensure the RESET, TMS, TCK, TDO, and TDI jumpers
+are in-place. Also place jumpers on the the TXD and RXD signals for a serial
```
change to ``in place`` (no hyphen)

dbkinder reviewed (changes requested):
Only one nit typo (and the missing images)

bwitherspoon commented:
I have not (yet) looked into BLE. The IEEE 802.15.4 driver is mostly complete, but not included in this PR.

I have fixed the typo in the documentation and added interrupt driven support to the UART driver.

galak commented on drivers/entropy/CMakeLists.txt:
```diff
@@ -7,6 +7,7 @@ zephyr_library_sources_ifdef(CONFIG_ENTROPY_MCUX_RNGA          entropy_mcux_rnga
 zephyr_library_sources_ifdef(CONFIG_ENTROPY_MCUX_TRNG          entropy_mcux_trng.c)
 zephyr_library_sources_ifdef(CONFIG_ENTROPY_NRF5_RNG           entropy_nrf5.c)
 zephyr_library_sources_ifdef(CONFIG_ENTROPY_SAM_RNG            entropy_sam.c)
+zephyr_library_sources_ifdef(CONFIG_ENTROPY_SIMPLELINK_RNG     entropy_simplelink.c)
```
Is the entropy driver needed for base functionality?  if not I'd ask to pull it into a separate PR after we get the base SoC/board ports merged.

bwitherspoon replied:
No it is not essential. I will split it into a seperate PR. I included it after reading [here](https://docs.zephyrproject.org/latest/reference/drivers/index.html#standard-drivers) that all supported boards have a random number generator driver, but I think I misunderstood.

galak commented on drivers/gpio/Kconfig:
```diff
@@ -62,6 +62,8 @@ source "drivers/gpio/Kconfig.sam0"
 
 source "drivers/gpio/Kconfig.sam"
 
+source "drivers/gpio/Kconfig.simplelink"
```
I think this should be called Kconfig.cc13x2_cc26x2

galak commented on drivers/gpio/Kconfig.simplelink:
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig GPIO_SIMPLELINK
```
GPIO_CC13X2_CC26X2

galak commented on drivers/pinmux/Kconfig.simplelink:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig PINMUX_SIMPLELINK
```
PINMUX_CC13X2_CC26X2

galak reviewed (changes requested):
Generally looks good, we need to rename simplelink to cc13x2_cc26x2 everywhere since simplelink is too generic a term to use.

bwitherspoon commented:
> Generally looks good, we need to rename simplelink to cc13x2_cc26x2 everywhere since simplelink is too generic a term to use.

I will rename. Originally I had named everything cc13x2_cc26x2 like the SoC, but realized the TI SDKs can be combined and the driver could also work on the older cc13x0 and cc26x0 series and possibly more of the SimpleLink family which have similar peripherals so I renamed it generically simplelink. But haven't been able to test this and that might have been overly optimistic.

bwitherspoon committed:
soc: ti_simplelink: add support for TI CC13x2 / CC26x2 series

Add initial support for the TI CC13x2 / CC26x2 series with the CC2652R
and CC1352R SoCs. The UART and GPIO peripherals are supported. Drivers
use the driverlib HAL from the TI CC13x2 / CC26x2 SDK.

Signed-off-by: Brett Witherspoon <spoonb@cdspooner.com>

bwitherspoon commented:
I have done the renaming, excluded the entropy driver, and added the images for the boards. I also fixed some of the license checks for files that are not in ext/ and from the SDK. I think all that is left from zephyrbot is the Kconfig whitelist and CODEOWNERS.

bwitherspoon commented:
I have confirmed these drivers work with the CC26x0 series (after adding the cc26x0 driverlib HAL) and presumably the CC13x0 too. Is it OK with the reviewers if I rename the pinmux, GPIO, and UART drivers cc13xx_cc26xx to reflect this? I have additional PRs for the CC26x0 series that will use these.

vanti commented on boards/arm/cc1352r_launchpad/support/openocd.cfg:
```diff
@@ -0,0 +1 @@
+source [find board/ti_cc26x2_launchpad.cfg]
```
It feels more appropriate to be using ti_cc13x2_launchpad.cfg instead for the cc1352r

vanti replied:
When I try to run 'make flash', the code loads and runs fine, but I am seeing this sort of error messages:

```

    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* cc13x2.cpu         cortex_m   little cc13x2.cpu         running
Info : JTAG tap: cc13x2.jrc tap/device found: 0x3bb4102f (mfg: 0x017 (Texas Instruments), part: 0xbb41, ver: 0x3)
Info : JTAG tap: cc13x2.cpu enabled
Error: Debug regions are unpowered, an unexpected reset might have happened
Error: JTAG-DP STICKY ERROR
Error: Could not find MEM-AP to control the core

```

Are you seeing the same? Just wondering if this is expected.

bwitherspoon replied:
Yes, I am seeing the same on both the CC2652 and the CC1352 boards. Not sure what the error is but loading, running and debugging appears fine.

vanti reviewed (commented):
Nice work. One general comment is for the board name used in directory, file, config names etc., it'd be nice for ‘cc1352r_launchpad’ to be renamed to ‘cc1352r_launchxl’ to be consistent with the naming used for CC3220SF. (TI SDKs tend to use ‘launchxl’ to refer to launchpads as well). Same applies to ‘cc26x2r_launchpad’. This way it'd be easier to remember what to type when switching board and running 'cmake .. -DBOARD=cc1352r_launchxl' for example.

GAnthony commented on soc/arm/ti_simplelink/cc13x2_cc26x2/Kconfig.series:
```diff
@@ -0,0 +1,15 @@
+config SOC_SERIES_CC13X2_CC26X2
+	bool "TI SimpleLink Family CC13x2 / CC26x2"
+	select CPU_CORTEX_M4
+	select SOC_FAMILY_TISIMPLELINK
```
If this is a CORTEX-M4F, with an FPU, wondering if we want to enable FPU and add:
`select CPU_HAS_FPU`
here?

bwitherspoon replied:
Yes, this is a Cortex-M4F so CPU_HAS_FPU makes sense. There is also an MPU so we might want to enable CPU_HAS_MPU.

GAnthony replied:
I'm guessing enabling FPU may save some footprint if doing any floating point.
Enabling MPU may have bigger consequences: may want to check the code to see what Zephyr will do/require in those cases.

bwitherspoon replied:
I'll let the MPU be enabled as needed in future PRs.

vanti commented on dts/arm/ti/cc1352r.dtsi:
```diff
@@ -0,0 +1,18 @@
+#include <ti/cc13x2_cc26x2.dtsi>
+
+/ {
+	sram0: memory@20000000 {
```
Looking at the hardware documentation, I see there are 8kb of cache memory on this device that can be optionally configured into RAM @0x11000000 by modifying ccfg.c and building it with the application. I think it needs to be defined as 'sram1'.

bwitherspoon commented:
> Nice work. One general comment is for the board name used in directory, file, config names etc., it'd be nice for ‘cc1352r_launchpad’ to be renamed to ‘cc1352r_launchxl’ to be consistent with the naming used for CC3220SF. (TI SDKs tend to use ‘launchxl’ to refer to launchpads as well). Same applies to ‘cc26x2r_launchpad’. This way it'd be easier to remember what to type when switching board and running 'cmake .. -DBOARD=cc1352r_launchxl' for example.

OK. The TI part numbers are LAUNCHXL-CC1352R1 and LAUNCHXL-CC26X2R1 so to be consistent with this, the TI SDK, and the CC3220SF-LAUNCHXL naming I renamed the boards 'cc1352r1_launchxl' and 'cc26x2r1_launchxl'. I left some of the informal 'TI CC26x2R LaunchPad' and 'TI CC1352R LaunchPad' in the documentation and Kconfig prompts, because I thought that might be helpful for someone looking for LaunchPad support. I hope that is not more confusing.

vanti commented:
> OK. The TI part numbers are LAUNCHXL-CC1352R1 and LAUNCHXL-CC26X2R1 so to be consistent with this, the TI SDK, and the CC3220SF-LAUNCHXL naming I renamed the boards 'cc1352r1_launchxl' and 'cc26x2r1_launchxl'. I left some of the informal 'TI CC26x2R LaunchPad' and 'TI CC1352R LaunchPad' in the documentation and Kconfig prompts, because I thought that might be helpful for someone looking for LaunchPad support. I hope that is not more confusing.

LGTM

bwitherspoon commented:
I believe I have resolved all reviewers comments. Also some (hopefully final) changes:

- Renamed the drivers cc13xx_cc26xx as mentioned in a previous comment.
- Enabling the FPU required adding some required CMSIS defines other than the defaults provided by Zephyr's cmsis.h. TI doesn't seem to provide a CMSIS device header, so they are in soc.h.
- Rebased on current master and updated for the CPU_HAS_SYSTICK to CPU_CORTEX_M_HAS_SYSTICK rename

Matt90o commented:
> 
> 
> I have not (yet) looked into BLE. The IEEE 802.15.4 driver is mostly complete, but not included in this PR.
> 
> I have fixed the typo in the documentation and added interrupt driven support to the UART driver.

Any place we could check out the IEEE 802.15.4 driver? Did you completely rewrite it or took TI's sources?

vanti reviewed (changes requested):
@bwitherspoon Thanks for addressing my comments. One more request: please split up the first commit such that the imported files appear in its own commit separate from your changes (the ones in the ccfg file for example). It makes it easier to see what has been modified when referring to it in the future.

bwitherspoon commented:
> > I have not (yet) looked into BLE. The IEEE 802.15.4 driver is mostly complete, but not included in this PR.
> > I have fixed the typo in the documentation and added interrupt driven support to the UART driver.
> 
> Any place we could check out the IEEE 802.15.4 driver? Did you completely rewrite it or took TI's sources?

I am currently working on pushing everything to github and adding draft PRs.

bwitherspoon commented:
> @bwitherspoon Thanks for addressing my comments. One more request: please split up the first commit such that the imported files appear in its own commit separate from your changes (the ones in the ccfg file for example). It makes it easier to see what has been modified when referring to it in the future.

No problem. Done.

bwitherspoon commented:
I came across a mistake in the UART driver due to a misunderstanding of how the device tree defines are generated. This would cause build failures when trying to enable UART 1 with UART 0. So the last force push fixes it to use the correct defines regardless of the UART port configuration. Also I used menuconfig when I should have used config for these options.

galak commented on ext/hal/ti/simplelink/source/ti/devices/cc13x2_cc26x2/startup_files/ccfg.c:
```diff
@@ -495,15 +498,8 @@
 // Customer Configuration Area in Lock Page
 //
 //*****************************************************************************
-#if defined(__IAR_SYSTEMS_ICC__)
```
I'd ask we keep the IAR & TI code here, and just handle the `#else` case

bwitherspoon replied:
OK. Fixed.

galak commented on ext/hal/ti/simplelink/README:
```diff
@@ -22,7 +22,8 @@ The driver library source is copied from the SDK, as follows:
 
 The source file startup_files/ccfg.c has been modified to use the appropriate
-sections when linking.
+sections when linking. In addition, the file inc/hw_flash.h has been modified
```
We should handle this by fixing/update the compliance script.

bwitherspoon replied:
I dropped that commit and will fix it in the compliance script. I originally had PR zephyrproject-rtos/ci-tools#52 to add CONFIG_CRC to the whitelist, but when I discovered it was just in a comment I took this approach instead. I can rebase and reopen that. It seems like the right solution might be for the tool to ignore comments.

galak reviewed (dismissed):
Generally looks good, a few minor things with related to the SDK import.  If you need help with the Kconfig CI check issue let me know.

bwitherspoon committed:
ext: simplelink: modify CC13x2/CC26x2 CCFG to use appropriate sections

The default customer configuration (CCFG) provided by the
TI CC13x2 / CC26x2 SDK puts the configuration in a section different
than expected by Zephyr. It has been modified to use the appropriate
section.

Signed-off-by: Brett Witherspoon <spoonb@cdspooner.com>

vanti reviewed (commented):
I tried to build the blinky sample for the cc1352r LaunchPad, and noticed that the -mfpu and -mfloat-abi compile options are not thrown:

> ccache /home/userlocal/bin/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBUILD_VERSION=zephyr-v1.14.0-315-g3df265fb49ec -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I/home/userlocal/zephyrproject-work/zephyr/kernel/include -I/home/userlocal/zephyrproject-work/zephyr/arch/arm/include -I/home/userlocal/zephyrproject-work/zephyr/include -I/home/userlocal/zephyrproject-work/zephyr/include/drivers -I/home/userlocal/zephyrproject-work/zephyr/samples/basic/blinky/build/zephyr/include/generated -I/home/userlocal/zephyrproject-work/zephyr/soc/arm/ti_simplelink/cc13x2_cc26x2 -I/home/userlocal/zephyrproject-work/zephyr/lib/libc/minimal/include -I/home/userlocal/zephyrproject-work/zephyr/ext/hal/cmsis/Include -I/home/userlocal/zephyrproject-work/zephyr/ext/hal/ti/simplelink/source/ti/devices/cc13x2_cc26x2/. -isystem /home/userlocal/bin/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/include -isystem /home/userlocal/bin/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/include-fixed  -Os -nostdinc -g -Wall -Wformat -Wformat-security -Wno-format-zero-length -imacros /home/userlocal/zephyrproject-work/zephyr/samples/basic/blinky/build/zephyr/include/generated/autoconf.h -ffreestanding -Wno-main -fno-common -mthumb -mcpu=cortex-m4 -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -Wno-pointer-sign -Wno-unused-but-set-variable -fno-reorder-functions -fno-defer-pop -Werror=implicit-int -Wpointer-arith -fmacro-prefix-map=/home/userlocal/zephyrproject-work/zephyr=. -ffunction-sections -fdata-sections -mabi=aapcs -march=armv7e-m -std=c99 -o CMakeFiles/app.dir/src/main.c.obj   -c /home/userlocal/zephyrproject-work/zephyr/samples/basic/blinky/src/main.c

These are options that are normally thrown in TI SDK examples, so I think something may be amiss in FPU support.

bwitherspoon commented:
> I tried to build the blinky sample for the cc1352r LaunchPad, and noticed that the -mfpu and -mfloat-abi compile options are not thrown:
> 
> > ccache /home/userlocal/bin/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBUILD_VERSION=zephyr-v1.14.0-315-g3df265fb49ec -DKERNEL -D_FORTIFY_SOURCE=2 -D__ZEPHYR__=1 -I/home/userlocal/zephyrproject-work/zephyr/kernel/include -I/home/userlocal/zephyrproject-work/zephyr/arch/arm/include -I/home/userlocal/zephyrproject-work/zephyr/include -I/home/userlocal/zephyrproject-work/zephyr/include/drivers -I/home/userlocal/zephyrproject-work/zephyr/samples/basic/blinky/build/zephyr/include/generated -I/home/userlocal/zephyrproject-work/zephyr/soc/arm/ti_simplelink/cc13x2_cc26x2 -I/home/userlocal/zephyrproject-work/zephyr/lib/libc/minimal/include -I/home/userlocal/zephyrproject-work/zephyr/ext/hal/cmsis/Include -I/home/userlocal/zephyrproject-work/zephyr/ext/hal/ti/simplelink/source/ti/devices/cc13x2_cc26x2/. -isystem /home/userlocal/bin/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/include -isystem /home/userlocal/bin/zephyr-sdk-0.10.0/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/8.3.0/include-fixed  -Os -nostdinc -g -Wall -Wformat -Wformat-security -Wno-format-zero-length -imacros /home/userlocal/zephyrproject-work/zephyr/samples/basic/blinky/build/zephyr/include/generated/autoconf.h -ffreestanding -Wno-main -fno-common -mthumb -mcpu=cortex-m4 -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -Wno-pointer-sign -Wno-unused-but-set-variable -fno-reorder-functions -fno-defer-pop -Werror=implicit-int -Wpointer-arith -fmacro-prefix-map=/home/userlocal/zephyrproject-work/zephyr=. -ffunction-sections -fdata-sections -mabi=aapcs -march=armv7e-m -std=c99 -o CMakeFiles/app.dir/src/main.c.obj   -c /home/userlocal/zephyrproject-work/zephyr/samples/basic/blinky/src/main.c
> 
> These are options that are normally thrown in TI SDK examples, so I think something may be amiss in FPU support.

FPU support can be enabled with `CONFIG_FLOAT=y` and then the you get the flags `-mfpu=fpv4-sp-d16 -mfloat-abi=hard` which should be correct. The documentation at https://docs.zephyrproject.org/latest/reference/kernel/other/float.html indicates the default mode is for no floating point registers to be used, so I think it is correct for it to not be enabled by default.

I set up a simple test project at https://github.com/bwitherspoon/fpu_test. Disassembling main (as described in README.rest) there are `vldr` and other floating point insturctions as expected with `-mfloat-abi=hard`. I did not test CONFIG_FP_SHARING.

vanti commented:
> 
> FPU support can be enabled with `CONFIG_FLOAT=y` and then you get the flags `-mfpu=fpv4-sp-d16 -mfloat-abi=hard` which should be correct. The documentation at https://docs.zephyrproject.org/latest/reference/kernel/other/float.html indicates the default mode is for no floating point registers to be used, so I think it is correct for it to not be enabled by default.
> 
> I set up a simple test project at https://github.com/bwitherspoon/fpu_test using `CONFIG_FLOAT=y` and tested on the CC1352 board. Disassembling main (as described in README.rst) there are `vldr` and other floating point instructions as expected with `-mfloat-abi=hard`. I did not test `CONFIG_FP_SHARING`.

Thanks for sharing. LGTM!

dbkinder reviewed (approved):
+1 for docs, thanks.

galak commented:
> drivers/gpio/Kconfig.cc13xx_cc26xx
> drivers/gpio/gpio_cc13xx_cc26xx.c
> drivers/pinmux/Kconfig.cc13xx_cc26xx
> drivers/pinmux/pinmux_cc13xx_cc26xx.c
> drivers/serial/Kconfig.cc13xx_cc26xx
> drivers/serial/uart_cc13xx_cc26xx.c
> dts/arm/ti/cc1352r.dtsi
> dts/arm/ti/cc13x2_cc26x2.dtsi
> dts/arm/ti/cc2652r.dtsi

Any plans to fix the CODEOWNERS?

If you don't feel like owning these you can ask @vanti to.

galak commented:
Merged the ci-tools/Kconfig CONFIG_CRC change so that check should hopefully pass now.

bwitherspoon committed:
CODEOWNERS: Add codeowner for TI CC13x2 / CC26x2 files

Add myself as codeowner for TI CC13x2 / CC26x2 related files.

Signed-off-by: Brett Witherspoon <spoonb@cdspooner.com>

bwitherspoon commented:
> > drivers/gpio/Kconfig.cc13xx_cc26xx
> > drivers/gpio/gpio_cc13xx_cc26xx.c
> > drivers/pinmux/Kconfig.cc13xx_cc26xx
> > drivers/pinmux/pinmux_cc13xx_cc26xx.c
> > drivers/serial/Kconfig.cc13xx_cc26xx
> > drivers/serial/uart_cc13xx_cc26xx.c
> > dts/arm/ti/cc1352r.dtsi
> > dts/arm/ti/cc13x2_cc26x2.dtsi
> > dts/arm/ti/cc2652r.dtsi
> 
> Any plans to fix the CODEOWNERS?
> 
> If you don't feel like owning these you can ask @vanti to.

I added myself to CODEOWNERS in a separate commit. I will squash it into another commit if necessary.

bwitherspoon commented:
> Any place we could check out the IEEE 802.15.4 driver? Did you completely rewrite it or took TI's sources?

@mfiumara See #16263
